[homematic ipwired] No function after a few days

@MHerbst  
after I have done the Update to 2.5.5 I have problems with Homematic IP Wired.

After rebooting openHab everything ist OK.
A few days later the HomematicWired things doesn't work.

Maybe there is a context to #7473 



![99289294_10221902480380948_8498322394115473408_o](https://user-images.githubusercontent.com/65848728/82750089-28a51800-9dae-11ea-8b4c-ba6ad66fb6ac.jpg)

> 2020-05-23 21:00:52.604 [ERROR] [ternal.handler.HomematicThingHandler] - null
> java.io.IOException: null
>         at org.openhab.binding.homematic.internal.communicator.client.XmlRpcClient.sendMessage(XmlRpcClient.java:116) ~[bundleFile:?]
>         at org.openhab.binding.homematic.internal.communicator.client.XmlRpcClient.sendMessage(XmlRpcClient.java:120) ~[bundleFile:?]
>         at org.openhab.binding.homematic.internal.communicator.client.XmlRpcClient.sendMessage(XmlRpcClient.java:71) ~[bundleFile:?]
>         at org.openhab.binding.homematic.internal.communicator.client.RpcClient.setDatapointValue(RpcClient.java:335) ~[bundleFile:?]
>         at org.openhab.binding.homematic.internal.communicator.AbstractHomematicGateway$1.execute(AbstractHomematicGateway.java:686) ~[bundleFile:?]
>         at org.openhab.binding.homematic.internal.misc.DelayedExecuter.start(DelayedExecuter.java:65) ~[bundleFile:?]
>         at org.openhab.binding.homematic.internal.communicator.AbstractHomematicGateway.sendDatapoint(AbstractHomematicGateway.java:665) ~[bundleFile:?]
>         at org.openhab.binding.homematic.internal.communicator.AbstractHomematicGateway.sendDatapoint(AbstractHomematicGateway.java:644) ~[bundleFile:?]
>         at org.openhab.binding.homematic.internal.handler.HomematicThingHandler.sendDatapoint(HomematicThingHandler.java:306) ~[bundleFile:?]
>         at org.openhab.binding.homematic.internal.handler.HomematicThingHandler.handleCommand(HomematicThingHandler.java:282) [bundleFile:?]
>         at sun.reflect.GeneratedMethodAccessor95.invoke(Unknown Source) ~[?:?]
>         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_252]
>         at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_252]
>         at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:152) [bundleFile:?]
>         at org.eclipse.smarthome.core.internal.common.InvocationHandlerSync.invoke(InvocationHandlerSync.java:59) [bundleFile:?]
>         at com.sun.proxy.$Proxy5493.handleCommand(Unknown Source) [?:?]
>         at org.eclipse.smarthome.core.thing.internal.profiles.ProfileCallbackImpl.handleCommand(ProfileCallbackImpl.java:74) [bundleFile:?]
>         at org.eclipse.smarthome.core.thing.internal.profiles.SystemDefaultProfile.onCommandFromItem(SystemDefaultProfile.java:48) [bundleFile:?]
>         at sun.reflect.GeneratedMethodAccessor94.invoke(Unknown Source) ~[?:?]
>         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_252]
>         at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_252]
>         at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:152) [bundleFile:?]
>         at org.eclipse.smarthome.core.internal.common.Invocation.call(Invocation.java:52) [bundleFile:?]
>         at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_252]
>         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_252]
>         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_252]
>         at java.lang.Thread.run(Thread.java:748) [?:1.8.0_252]
> Caused by: java.util.concurrent.TimeoutException
>         at org.eclipse.jetty.client.util.InputStreamResponseListener.get(InputStreamResponseListener.java:216) ~[bundleFile:9.4.20.v20190813]
>         at org.openhab.binding.homematic.internal.communicator.client.XmlRpcClient.sendMessage(XmlRpcClient.java:89) ~[bundleFile:?]
>         ... 26 more

i get this every 30 seconds after the error appeared for the first time

crashed again, here your new logging (every 30 seconds from this time on):

> 2020-05-25 22:12:25.706 [ERROR] [ternal.handler.HomematicThingHandler] - setValuefailed: null
> java.io.IOException: setValuefailed: null
>         at org.openhab.binding.homematic.internal.communicator.client.XmlRpcClient.sendMessage(XmlRpcClient.java:117) ~[bundleFile:?]
>         at org.openhab.binding.homematic.internal.communicator.client.XmlRpcClient.sendMessage(XmlRpcClient.java:121) ~[bundleFile:?]
>         at org.openhab.binding.homematic.internal.communicator.client.XmlRpcClient.sendMessage(XmlRpcClient.java:71) ~[bundleFile:?]
>         at org.openhab.binding.homematic.internal.communicator.client.RpcClient.setDatapointValue(RpcClient.java:335) ~[bundleFile:?]
>         at org.openhab.binding.homematic.internal.communicator.AbstractHomematicGateway$1.execute(AbstractHomematicGateway.java:686) ~[bundleFile:?]
>         at org.openhab.binding.homematic.internal.misc.DelayedExecuter.start(DelayedExecuter.java:65) ~[bundleFile:?]
>         at org.openhab.binding.homematic.internal.communicator.AbstractHomematicGateway.sendDatapoint(AbstractHomematicGateway.java:665) ~[bundleFile:?]
>         at org.openhab.binding.homematic.internal.communicator.AbstractHomematicGateway.sendDatapoint(AbstractHomematicGateway.java:644) ~[bundleFile:?]
>         at org.openhab.binding.homematic.internal.handler.HomematicThingHandler.sendDatapoint(HomematicThingHandler.java:306) ~[bundleFile:?]
>         at org.openhab.binding.homematic.internal.handler.HomematicThingHandler.handleCommand(HomematicThingHandler.java:282) [bundleFile:?]
>         at sun.reflect.GeneratedMethodAccessor126.invoke(Unknown Source) ~[?:?]
>         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_252]
>         at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_252]
>         at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:152) [bundleFile:?]
>         at org.eclipse.smarthome.core.internal.common.InvocationHandlerSync.invoke(InvocationHandlerSync.java:59) [bundleFile:?]
>         at com.sun.proxy.$Proxy6134.handleCommand(Unknown Source) [?:?]
>         at org.eclipse.smarthome.core.thing.internal.profiles.ProfileCallbackImpl.handleCommand(ProfileCallbackImpl.java:74) [bundleFile:?]
>         at org.eclipse.smarthome.core.thing.internal.profiles.SystemDefaultProfile.onCommandFromItem(SystemDefaultProfile.java:48) [bundleFile:?]
>         at sun.reflect.GeneratedMethodAccessor125.invoke(Unknown Source) ~[?:?]
>         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_252]
>         at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_252]
>         at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:152) [bundleFile:?]
>         at org.eclipse.smarthome.core.internal.common.Invocation.call(Invocation.java:52) [bundleFile:?]
>         at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_252]
>         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_252]
>         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_252]
>         at java.lang.Thread.run(Thread.java:748) [?:1.8.0_252]
> Caused by: java.util.concurrent.TimeoutException
>         at org.eclipse.jetty.client.util.InputStreamResponseListener.get(InputStreamResponseListener.java:216) ~[bundleFile:9.4.20.v20190813]
>         at org.openhab.binding.homematic.internal.communicator.client.XmlRpcClient.sendMessage(XmlRpcClient.java:89) ~[bundleFile:?]
>         ... 26 more
> 2020-05-25 22:13:12.332 [INFO ] [nal.communicator.client.XmlRpcClient] - XmlRpcMessage failure, sending message again 1/1
> 

and as before, if i do a:

> ssh openhab@localhost -p8101 'bundle:restart org.openhab.binding.homematic'

i get:

> [INFO ] [ommunicator.AbstractHomematicGateway] - HmGatewayInfo[id=CCU,type=CCU2,firmware=3.51.6.20200420,address=PEQ1944622,rf=true,wired=false,hmip=false,cuxd=false,group=true]

(hmip=false)
so something in openHAB seems to be crashed/hung at the time the error appeared first

@MHerbst  since today I have problemens with Homematic.

Here the Log
2020-05-27 18:53:12.418 [hingStatusInfoChangedEvent] - 'homematic:HM-LC-Bl1PBU-FM:3014F711A0001F9A498BA50A:OEQ0886178' changed from ONLINE to OFFLINE (COMMUNICATION_ERROR)

==> /var/log/openhab2/openhab.log <==

2020-05-27 18:53:12.412 [ERROR] [ternal.handler.HomematicThingHandler] - -1 Failure (sending <?xml version="1.0" encoding="ISO-8859-1"?>

<methodCall><methodName>setValue</methodName>

<params><param><value>OEQ0886178:1</value></param><param><value>LEVEL</value></param><param><value><double>0.99</double></value></param></params></methodCall>)

org.openhab.binding.homematic.internal.communicator.client.UnknownRpcFailureException: -1 Failure (sending <?xml version="1.0" encoding="ISO-8859-1"?>

<methodCall><methodName>setValue</methodName>

<params><param><value>OEQ0886178:1</value></param><param><value>LEVEL</value></param><param><value><double>0.99</double></value></param></params></methodCall>)

	at org.openhab.binding.homematic.internal.communicator.parser.RpcResponseParser.parse(RpcResponseParser.java:47) ~[bundleFile:?]

	at org.openhab.binding.homematic.internal.communicator.client.XmlRpcClient.sendMessage(XmlRpcClient.java:112) ~[bundleFile:?]

	at org.openhab.binding.homematic.internal.communicator.client.XmlRpcClient.sendMessage(XmlRpcClient.java:71) ~[bundleFile:?]

	at org.openhab.binding.homematic.internal.communicator.client.RpcClient.setDatapointValue(RpcClient.java:335) ~[bundleFile:?]

	at org.openhab.binding.homematic.internal.communicator.AbstractHomematicGateway$1.execute(AbstractHomematicGateway.java:686) ~[bundleFile:?]

	at org.openhab.binding.homematic.internal.misc.DelayedExecuter.start(DelayedExecuter.java:65) ~[bundleFile:?]

	at org.openhab.binding.homematic.internal.communicator.AbstractHomematicGateway.sendDatapoint(AbstractHomematicGateway.java:665) ~[bundleFile:?]

	at org.openhab.binding.homematic.internal.communicator.AbstractHomematicGateway.sendDatapoint(AbstractHomematicGateway.java:644) ~[bundleFile:?]

	at org.openhab.binding.homematic.internal.handler.HomematicThingHandler.sendDatapoint(HomematicThingHandler.java:306) ~[bundleFile:?]

	at org.openhab.binding.homematic.internal.handler.HomematicThingHandler.handleCommand(HomematicThingHandler.java:282) [bundleFile:?]

	at sun.reflect.GeneratedMethodAccessor817.invoke(Unknown Source) ~[?:?]

	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_252]

	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_252]

	at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:152) [bundleFile:?]

	at org.eclipse.smarthome.core.internal.common.InvocationHandlerSync.invoke(InvocationHandlerSync.java:59) [bundleFile:?]

	at com.sun.proxy.$Proxy22828.handleCommand(Unknown Source) [?:?]

	at org.eclipse.smarthome.core.thing.internal.profiles.ProfileCallbackImpl.handleCommand(ProfileCallbackImpl.java:74) [bundleFile:?]

	at org.eclipse.smarthome.core.thing.internal.profiles.SystemDefaultProfile.onCommandFromItem(SystemDefaultProfile.java:48) [bundleFile:?]

	at sun.reflect.GeneratedMethodAccessor816.invoke(Unknown Source) ~[?:?]

	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_252]

	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_252]

	at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:152) [bundleFile:?]

	at org.eclipse.smarthome.core.internal.common.Invocation.call(Invocation.java:52) [bundleFile:?]

	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_252]

	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_252]

	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_252]

	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_252]

ok, thanks. i restarted twice as everytime after cleaning the cache and at the second restart there was no error. 

the binding runs for nearly 24 hours - it had 3 error messages like this:

```
2020-06-02 15:01:45.748 [ERROR] [ternal.handler.HomematicThingHandler] - Total timeout 15000 ms elapsed
java.io.IOException: Total timeout 15000 ms elapsed
        at org.openhab.binding.homematic.internal.communicator.client.XmlRpcClient.sendMessage(XmlRpcClient.java:83) ~[bundleFile:?]
        at org.openhab.binding.homematic.internal.communicator.client.RpcClient.setDatapointValue(RpcClient.java:335) ~[bundleFile:?]
        at org.openhab.binding.homematic.internal.communicator.AbstractHomematicGateway$1.execute(AbstractHomematicGateway.java:686) ~[bundleFile:?]
        at org.openhab.binding.homematic.internal.misc.DelayedExecuter.start(DelayedExecuter.java:65) ~[bundleFile:?]
        at org.openhab.binding.homematic.internal.communicator.AbstractHomematicGateway.sendDatapoint(AbstractHomematicGateway.java:665) ~[bundleFile:?]
        at org.openhab.binding.homematic.internal.communicator.AbstractHomematicGateway.sendDatapoint(AbstractHomematicGateway.java:644) ~[bundleFile:?]
        at org.openhab.binding.homematic.internal.handler.HomematicThingHandler.sendDatapoint(HomematicThingHandler.java:306) ~[bundleFile:?]
        at org.openhab.binding.homematic.internal.handler.HomematicThingHandler.handleCommand(HomematicThingHandler.java:282) [bundleFile:?]
        at sun.reflect.GeneratedMethodAccessor105.invoke(Unknown Source) ~[?:?]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_252]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_252]
        at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:152) [bundleFile:?]
        at org.eclipse.smarthome.core.internal.common.InvocationHandlerSync.invoke(InvocationHandlerSync.java:59) [bundleFile:?]
        at com.sun.proxy.$Proxy2911.handleCommand(Unknown Source) [?:?]
        at org.eclipse.smarthome.core.thing.internal.profiles.ProfileCallbackImpl.handleCommand(ProfileCallbackImpl.java:74) [bundleFile:?]
        at org.eclipse.smarthome.core.thing.internal.profiles.SystemDefaultProfile.onCommandFromItem(SystemDefaultProfile.java:48) [bundleFile:?]
        at sun.reflect.GeneratedMethodAccessor104.invoke(Unknown Source) ~[?:?]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_252]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_252]
        at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:152) [bundleFile:?]
        at org.eclipse.smarthome.core.internal.common.Invocation.call(Invocation.java:52) [bundleFile:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_252]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_252]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_252]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_252]
Caused by: java.util.concurrent.TimeoutException: Total timeout 15000 ms elapsed
        at org.eclipse.jetty.client.TimeoutCompleteListener.onTimeoutExpired(TimeoutCompleteListener.java:51) ~[?:?]
        at org.eclipse.jetty.io.CyclicTimeout$Wakeup.run(CyclicTimeout.java:282) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_252]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_252]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) ~[?:1.8.0_252]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) ~[?:1.8.0_252]
        ... 3 more

```

But everything is still working.

--- a/bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/client/RpcClient.java
+++ b/bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/client/RpcClient.java
@@ -51,7 +51,7 @@ import org.slf4j.LoggerFactory;
  */
 public abstract class RpcClient<T> {
     private final Logger logger = LoggerFactory.getLogger(RpcClient.class);
-    protected static final int MAX_RPC_RETRY = 1;
+    protected static final int MAX_RPC_RETRY = 3;
     protected static final int RESP_BUFFER_SIZE = 8192;
 
     protected HomematicConfig config;

--- a/bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/client/XmlRpcClient.java
+++ b/bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/client/XmlRpcClient.java
@@ -16,7 +16,12 @@ import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
+import java.io.UnsupportedEncodingException;
+import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import javax.xml.parsers.ParserConfigurationException;
 
 import org.eclipse.jetty.client.HttpClient;
 import org.eclipse.jetty.client.api.Request;
@@ -32,6 +37,7 @@ import org.openhab.binding.homematic.internal.communicator.message.XmlRpcRespons
 import org.openhab.binding.homematic.internal.communicator.parser.RpcResponseParser;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import org.xml.sax.SAXException;
 
 /**
  * Client implementation for sending messages via XML-RPC to the Homematic server.
@@ -69,7 +75,7 @@ public class XmlRpcClient extends RpcClient<String> {
         if (logger.isTraceEnabled()) {
             logger.trace("Client XmlRpcRequest (port {}):\n{}", port, request);
         }
-        Exception reason = null;
+        IOException reason = new IOException();
         for (int rpcRetryCounter = 1; rpcRetryCounter <= MAX_RPC_RETRY; rpcRetryCounter++) {
             try {
                 byte[] response = send(port, request);
@@ -78,57 +84,64 @@ public class XmlRpcClient extends RpcClient<String> {
                 return new RpcResponseParser(request).parse(data);
             } catch (UnknownRpcFailureException | UnknownParameterSetException ex) {
                 throw ex;
-            } catch (Exception ex) {
+            } catch (SAXException | ParserConfigurationException ex) {
+                throw new IOException(ex);
+            } catch (IOException ex) {
                 reason = ex;
                 if ("init".equals(request.getMethodName())) { // no retries for "init" request
                     break;
                 }
-                logger.debug("XmlRpcMessage failure, sending message again {}/{}", rpcRetryCounter, MAX_RPC_RETRY);
+                logger.debug("XmlRpcMessage failed, sending message again {}/{}", rpcRetryCounter, MAX_RPC_RETRY);
             }
         }
-        throw new IOException(reason.getMessage(), reason); // can't be null here because logic in for loop
+        throw reason;
     }
 
-    private byte[] send(int port, RpcRequest<String> request) throws Exception {
+    private byte[] send(int port, RpcRequest<String> request) throws IOException {
         byte[] ret = new byte[0];
-        BytesContentProvider content = new BytesContentProvider(request.createMessage().getBytes(config.getEncoding()));
-        String url = String.format("http://%s:%s", config.getGatewayAddress(), port);
-        if (port == config.getGroupPort()) {
-            url += "/groups";
-        }
-        Request req = httpClient.POST(url).content(content).timeout(config.getTimeout(), TimeUnit.SECONDS)
-                .header(HttpHeader.CONTENT_TYPE, "text/xml;charset=" + config.getEncoding());
         try {
-            ret = req.send().getContent();
-        } catch (IllegalArgumentException e) { // Returned buffer too large
-            logger.info("Blocking XmlRpcRequest failed: {}, trying non-blocking request", e.getMessage());
-            InputStreamResponseListener respListener = new InputStreamResponseListener();
-            req.send(respListener);
-            Response resp = respListener.get(config.getTimeout(), TimeUnit.SECONDS);
-            ByteArrayOutputStream respData = new ByteArrayOutputStream(RESP_BUFFER_SIZE);
-
-            int httpStatus = resp.getStatus();
-            if (httpStatus == HttpStatus.OK_200) {
-                byte[] recvBuffer = new byte[RESP_BUFFER_SIZE];
-                try (InputStream input = respListener.getInputStream()) {
-                    while (true) {
-                        int read = input.read(recvBuffer);
-                        if (read == -1) {
-                            break;
-                        }
-                        respData.write(recvBuffer, 0, read);
-                    }
-                    ret = respData.toByteArray();
-                }
-            } else {
-                logger.warn("Non-blocking XmlRpcRequest failed, status code: {} / request: {}", httpStatus, request);
-                resp.abort(new Exception());
+            BytesContentProvider content = new BytesContentProvider(
+                    request.createMessage().getBytes(config.getEncoding()));
+            String url = String.format("http://%s:%s", config.getGatewayAddress(), port);
+            if (port == config.getGroupPort()) {
+                url += "/groups";
             }
-        }
+            Request req = httpClient.POST(url).content(content).timeout(config.getTimeout(), TimeUnit.SECONDS)
+                    .header(HttpHeader.CONTENT_TYPE, "text/xml;charset=" + config.getEncoding());
+            try {
+                ret = req.send().getContent();
+            } catch (IllegalArgumentException e) { // Returned buffer too large
+                logger.info("Blocking XmlRpcRequest failed: {}, trying non-blocking request", e.getMessage());
+                InputStreamResponseListener respListener = new InputStreamResponseListener();
+                req.send(respListener);
+                Response resp = respListener.get(config.getTimeout(), TimeUnit.SECONDS);
+                ByteArrayOutputStream respData = new ByteArrayOutputStream(RESP_BUFFER_SIZE);
 
-        if (logger.isTraceEnabled()) {
-            String result = new String(ret, config.getEncoding());
-            logger.trace("Client XmlRpcResponse (port {}):\n{}", port, result);
+                int httpStatus = resp.getStatus();
+                if (httpStatus == HttpStatus.OK_200) {
+                    byte[] recvBuffer = new byte[RESP_BUFFER_SIZE];
+                    try (InputStream input = respListener.getInputStream()) {
+                        while (true) {
+                            int read = input.read(recvBuffer);
+                            if (read == -1) {
+                                break;
+                            }
+                            respData.write(recvBuffer, 0, read);
+                        }
+                        ret = respData.toByteArray();
+                    }
+                } else {
+                    logger.warn("Non-blocking XmlRpcRequest failed, status code: {} / request: {}", httpStatus,
+                            request);
+                    resp.abort(new Exception());
+                }
+            }
+            if (logger.isTraceEnabled()) {
+                String result = new String(ret, config.getEncoding());
+                logger.trace("Client XmlRpcResponse (port {}):\n{}", port, result);
+            }
+        } catch (UnsupportedEncodingException | ExecutionException | TimeoutException | InterruptedException e) {
+            throw new IOException(e);
         }
         return ret;
     }

--- a/bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/client/XmlRpcClient.java
+++ b/bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/client/XmlRpcClient.java
@@ -133,7 +133,7 @@ public class XmlRpcClient extends RpcClient<String> {
                 } else {
                     logger.warn("Non-blocking XmlRpcRequest failed, status code: {} / request: {}", httpStatus,
                             request);
-                    resp.abort(new Exception());
+                    resp.abort(new IOException());
                 }
             }
             if (logger.isTraceEnabled()) {

--- a/bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/client/XmlRpcClient.java
+++ b/bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/client/XmlRpcClient.java
@@ -19,6 +19,7 @@ import java.io.InputStream;
 import java.util.concurrent.TimeUnit;
 
 import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.api.Request;
 import org.eclipse.jetty.client.api.Response;
 import org.eclipse.jetty.client.util.BytesContentProvider;
 import org.eclipse.jetty.client.util.InputStreamResponseListener;
@@ -68,57 +69,67 @@ public class XmlRpcClient extends RpcClient<String> {
         if (logger.isTraceEnabled()) {
             logger.trace("Client XmlRpcRequest (port {}):\n{}", port, request);
         }
-        return sendMessage(port, request, 0);
+        Exception reason = null;
+        for (int rpcRetryCounter = 1; rpcRetryCounter <= MAX_RPC_RETRY; rpcRetryCounter++) {
+            try {
+                byte[] response = send(port, request);
+                Object[] data = new XmlRpcResponse(new ByteArrayInputStream(response), config.getEncoding())
+                        .getResponseData();
+                return new RpcResponseParser(request).parse(data);
+            } catch (UnknownRpcFailureException | UnknownParameterSetException ex) {
+                throw ex;
+            } catch (Exception ex) {
+                reason = ex;
+                if ("init".equals(request.getMethodName())) { // no retries for "init" request
+                    break;
+                }
+                logger.debug("XmlRpcMessage failure, sending message again {}/{}", rpcRetryCounter, MAX_RPC_RETRY);
+            }
+        }
+        throw new IOException(reason.getMessage(), reason); // can't be null here because logic in for loop
     }
 
-    /**
-     * Sends the message, retries if there was an error.
-     */
-    private synchronized Object[] sendMessage(int port, RpcRequest<String> request, int rpcRetryCounter)
-            throws IOException {
+    private byte[] send(int port, RpcRequest<String> request) throws Exception {
+        byte[] ret = new byte[0];
+        BytesContentProvider content = new BytesContentProvider(request.createMessage().getBytes(config.getEncoding()));
+        String url = String.format("http://%s:%s", config.getGatewayAddress(), port);
+        if (port == config.getGroupPort()) {
+            url += "/groups";
+        }
+        Request req = httpClient.POST(url).content(content).timeout(config.getTimeout(), TimeUnit.SECONDS)
+                .header(HttpHeader.CONTENT_TYPE, "text/xml;charset=" + config.getEncoding());
         try {
-            BytesContentProvider content = new BytesContentProvider(
-                    request.createMessage().getBytes(config.getEncoding()));
-            String url = String.format("http://%s:%s", config.getGatewayAddress(), port);
-            if (port == config.getGroupPort()) {
-                url += "/groups";
-            }
+            ret = req.send().getContent();
+        } catch (IllegalArgumentException e) { // Returned buffer too large
+            logger.info("Blocking XmlRpcRequest failed: {}, trying non-blocking request", e.getMessage());
             InputStreamResponseListener respListener = new InputStreamResponseListener();
-            httpClient.POST(url).content(content)
-                    .header(HttpHeader.CONTENT_TYPE, "text/xml;charset=" + config.getEncoding()).send(respListener);
+            req.send(respListener);
             Response resp = respListener.get(config.getTimeout(), TimeUnit.SECONDS);
             ByteArrayOutputStream respData = new ByteArrayOutputStream(RESP_BUFFER_SIZE);
+
             int httpStatus = resp.getStatus();
             if (httpStatus == HttpStatus.OK_200) {
                 byte[] recvBuffer = new byte[RESP_BUFFER_SIZE];
-                InputStream input = respListener.getInputStream();
-                while (true) {
-                    int read = input.read(recvBuffer);
-                    if (read == -1) {
-                        break;
+                try (InputStream input = respListener.getInputStream()) {
+                    while (true) {
+                        int read = input.read(recvBuffer);
+                        if (read == -1) {
+                            break;
+                        }
+                        respData.write(recvBuffer, 0, read);
                     }
-                    respData.write(recvBuffer, 0, read);
+                    ret = respData.toByteArray();
                 }
             } else {
-                logger.warn("XmlRpcRequest failure, status code: {} / request was: {}", httpStatus, request);
+                logger.warn("Non-blocking XmlRpcRequest failed, status code: {} / request: {}", httpStatus, request);
                 resp.abort(new Exception());
             }
-            if (logger.isTraceEnabled()) {
-                logger.trace("Client XmlRpcResponse: (port {}):\n{}", port, respData.toString(config.getEncoding()));
-            }
-            Object[] data = new XmlRpcResponse(new ByteArrayInputStream(respData.toByteArray()), config.getEncoding())
-                    .getResponseData();
-            return new RpcResponseParser(request).parse(data);
-        } catch (UnknownRpcFailureException | UnknownParameterSetException ex) {
-            throw ex;
-        } catch (Exception ex) {
-            if ("init".equals(request.getMethodName()) || rpcRetryCounter >= MAX_RPC_RETRY) {
-                throw new IOException(ex.getMessage(), ex);
-            } else {
-                rpcRetryCounter++;
-                logger.debug("XmlRpcMessage failure, sending message again {}/{}", rpcRetryCounter, MAX_RPC_RETRY);
-                return sendMessage(port, request, rpcRetryCounter);
-            }
         }
+
+        if (logger.isTraceEnabled()) {
+            String result = new String(ret, config.getEncoding());
+            logger.trace("Client XmlRpcResponse (port {}):\n{}", port, result);
+        }
+        return ret;
     }
 }

